cryptoki/src/flag.rs

```diff
@@ -0,0 +1,319 @@
+// Copyright 2021 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Generic abstraction of CK_FLAGS for specialization in other modules
+//!
+//! Generate a bit flag set bound to a type definition, T. T has no
+//! function beyond providing a compile-time bounds check and is only
+//! present as phantom data.
+//!
+//! These definitions also divide the roles of a flag accumulator
+//! (`BitFlags`) from a single, constant bit that contributes to it
+//! ('FlagBit`) or that bit's inverse (`FlagMask`)
+//!
+//! The following fundamental operations are defined.
+//!
+//! * Flags traits:  From CK_FLAGS, Deref and DerefMut as CK_FLAGS
+//! * !Bit -> Mask (ones compliment)
+//! * Flags |= Bit
+//! * Flags &= Mask
+//! * Flags ^= Bit
+//! * Bit | Bit
+//! * Mask & Mask
+//! * Flags.contains(Bit)
+//!
+//! Imporantly, Bit types cannot* be assigned to, but can be composed
+//! for assignment to the associated accumulator type. Likewise, two
+//! accumulators can't be composed with each other. Only the accumulator
+//! type can be converted to/from CK_FLAGS for use in FFI operations.
+//!
+//! *Technically false. The binary operations resolve to a bit type, not
+//! an accumulator. If assigned to an itermediate variable, a multi-bit
+//! type can be constructed. There is also no enforcement of the assumption
+//! that FlagBit only has a single set bit. The lack of assign-op operators
+//! is intended to discourage this, though, and the type isn't pub to be
+//! abused anyway.
+//!
+//! The following are intended use patterns.
+//!
+//! * Set: Flags |= Bit
+//! * Set multiple: Flags |= Bit | Bit | ...
+//! * Unset: Flags &= !Bit
+//! * Unset multiple: Flags &= !(Bit | Bit | ...)
+//! * Toggle: Flags ^= Bit
+//! * Toggle mutliple: Flags ^= Bit | Bit | ...
+//! * Test: Flags.contains(Bit) -> bool
+//! * Test (ALL): Flags.contains(Bit | Bit | ...) -> bool
+use cryptoki_sys::CK_FLAGS;
+use std::marker::PhantomData;
+use std::ops::{BitAnd, BitAndAssign, BitOr, BitOrAssign, BitXorAssign, Deref, DerefMut, Not};
+
+/// A type representing a 1-bit flag constant or some composition thereof
+#[repr(transparent)]
+pub(crate) struct FlagBit<T> {
+    value: CK_FLAGS,
+    _type: PhantomData<T>,
+}
+
+/// A type representing the one's compliment of a FlagBit<T>
+#[repr(transparent)]
+pub(crate) struct FlagMask<T> {
```

Comment:
This is a very nice implementation of bit fields & masks. This is mostly generic - I was wondering - isn't there a crate the we could reuse for this purpose?

Reply:
Yes! There is a [bitflags](https://docs.rs/bitflags/latest/bitflags/) crate, which I may actually switch over to before calling this done. The reason I avoided it initially, is I thought these types would be public, and so wanted them bound to their containing struct to avoid any kind of confusion or misuse allowed by the public API. But once I got to the struct itself, I found that they could easily be private. So don't be surprised if I do end up doing exactly that.